src/notation/view/notationviewinputcontroller.cpp

```diff
@@ -657,135 +653,199 @@ void NotationViewInputController::mousePressEvent(QMouseEvent* event)
+        return;
     }
 
+    // Misc
     if (button == Qt::LeftButton) {
         handleLeftClick(ctx);
     } else if (button == Qt::RightButton) {
         handleRightClick(ctx);
     }
 }
 
-bool NotationViewInputController::needSelect(const ClickContext& ctx) const
+void NotationViewInputController::handleClickInNoteInputMode(QMouseEvent* event)
+{
+    const PointF logicPos = m_view->toLogical(event->pos());
+
+    if (event->button() == Qt::RightButton) {
+        dispatcher()->dispatch("remove-note", ActionData::make_arg1<PointF>(logicPos));
+        return;
+    }
+
+    const Qt::KeyboardModifiers keyState = event->modifiers();
+    const bool replace = keyState & Qt::ShiftModifier;
+    const bool insert = keyState & Qt::ControlModifier;
+    dispatcher()->dispatch("put-note", ActionData::make_arg3<PointF, bool, bool>(logicPos, replace, insert));
+}
+
+bool NotationViewInputController::tryHandleClickGrip(const ClickContext& ctx)
 {
     if (ctx.isHitGrip && ctx.event->button() == Qt::LeftButton) {
+        viewInteraction()->startEditGrip(ctx.logicClickPos);
+        return true;
+    } else if (viewInteraction()->isGripEditStarted()) {
+        viewInteraction()->endEditGrip();
+    }
+
+    return false;
+}
+
+bool NotationViewInputController::tryHandleClickDragOutgoingElement(const ClickContext& ctx)
+{
+    if (ctx.event->button() != Qt::LeftButton
+        || ctx.event->modifiers() != (Qt::ShiftModifier | Qt::ControlModifier)) {
         return false;
     }
 
-    if (!ctx.hitElement) {
+    if (!ctx.hitElement || !viewInteraction()->isOutgoingDragElementAllowed(ctx.hitElement)) {
         return false;
     }
 
+    viewInteraction()->select({ ctx.hitElement }, SelectType::SINGLE, ctx.hitStaff);
+    m_mouseDownInfo.dragAction = MouseDownInfo::DragOutgoingElement;
+    return true;
+}
+
+void NotationViewInputController::handleClickSelect(const ClickContext& ctx)
+{
+    if (!ctx.hitElement) {
+        return;
+    }
+
     INotationSelectionPtr selection = viewInteraction()->selection();
 
-    if (ctx.event->button() == Qt::LeftButton && ctx.event->modifiers() & Qt::ControlModifier) {
-        return true;
-    } else if (ctx.event->button() == Qt::LeftButton && ctx.event->modifiers() & Qt::ShiftModifier) {
-        return !selection->isRange() || !selection->range()->containsItem(ctx.hitElement, ctx.hitStaff);
-    } else if (ctx.event->button() == Qt::RightButton && selection->isRange()) {
-        return !selection->range()->containsItem(ctx.hitElement, ctx.hitStaff);
-    } else if (!ctx.hitElement->selected()) {
-        return true;
+    if (ctx.event->button() == Qt::LeftButton) {
+        if (ctx.event->modifiers() & Qt::ControlModifier) {
+            std::vector<EngravingItem*> overlappingHitElements = viewInteraction()->hitElements(ctx.logicClickPos, hitWidth());
+            if (overlappingHitElements.size() > 1) {
+                cycleThroughOverlappingHitElements(overlappingHitElements, ctx.hitStaff);
```

Comment:
Couple of nitpicks here:

```suggestion
            const std::vector<EngravingItem*> overlappingHitElements = viewInteraction()->hitElements(ctx.logicClickPos, hitWidth());
            if (overlappingHitElements.size() > 1) {
                cycleOverlappingHitElements(overlappingHitElements, ctx.hitStaff);
```

---

There's also perhaps an argument that we could do something like this, I'll leave it up to you:
```
            const bool didCycle = cycleOverlappingAtPosition(ctx.logicClickPos, ctx.hitStaff);
            if (didCycle) {
                return;
            }
```